src/query/queryWithCursor.ts

```diff
@@ -0,0 +1,50 @@
+import { DocumentClient, Key, QueryOutput } from 'aws-sdk/clients/dynamodb';
+import { decrypt, encrypt } from '../utils';
+import { AnyObject, Filter, TableConfig } from '../types';
+import { buildQueryTableParams } from './queryBuilder';
+
+/**
+ * Cursor based query which returns list of matching items and the last cursor position
+ *
+ * @param {Filter<T>} filter query filter
+ * @param { QueryWithCursorOptions } options cursor, page size options
+ * @returns {Array<T>, string} list of matching items, last cursor position
+ */
+export async function queryWithCursor<T extends AnyObject>(
+  dbClient: DocumentClient,
+  table: TableConfig,
+  filter: Filter<T>,
+  indexName = 'default',
+): Promise<{ items: Array<T>; cursor?: string }> {
+  const { partitionKeyName, sortKeyName } = table.indexes[indexName];
+
+  if (!sortKeyName) {
+    throw new Error('Expected sortKey to query');
+  }
+
+  if (!table.cursorSecret) {
+    throw new Error(
+      'Expected `cursorSecret` which is used to encrypt the `LastEvaluatedKey`',
+    );
+  }
+
+  const params = buildQueryTableParams(filter, partitionKeyName, sortKeyName);
+
+  params.TableName = table.name;
+  if (indexName) {
+    params.IndexName = indexName;
+  }
+  params.ExclusiveStartKey = decrypt<Key>(
```

Comment:
We should be able to handle the scenario where there is no cursor key passed for the first call?

Reply:
Yes it was handled.
The `decrypt` will return `undefined` when `prevCursor` is `empty`!. 
When the `ExclusiveStartKey` is empty it will query from the first record (where records are sorted in ascending - default sort mechanism by **DynamoDB**)